File: scripts/generate-icons.py
"""Generate PWA icon PNGs (192x192 and 512x512) using only stdlib."""
import struct
import zlib
import os

OUT_DIR = os.path.join(os.path.dirname(__file__), "..", "public")

def create_png(width: int, height: int) -> bytes:
    """Create a minimal PNG with a simple '票影' themed icon design."""
    # BG: warm cream #F7F3EA, accent: dark brown #5D4037
    BG_R, BG_G, BG_B = 0xF7, 0xF3, 0xEA
    ACC_R, ACC_G, ACC_B = 0x5D, 0x40, 0x37

    # Build raw pixel rows (RGBA)
    raw = b""
    cx, cy = width // 2, height // 2
    radius = int(width * 0.38)

    for y in range(height):
        raw += b"\x00"  # filter byte: none
        for x in range(width):
            dx, dy = x - cx, y - cy
            dist = (dx * dx + dy * dy) ** 0.5

            # Rounded square with circle cutout
            in_circle = dist <= radius
            in_square = (abs(dx) <= radius and abs(dy) <= radius)

            # Draw a document shape: rounded rect with folded corner
            half = width // 2
            qr = width // 4
            # Simple design: circle on cream background
            if in_circle:
                r, g, b, a = ACC_R, ACC_G, ACC_B, 255
            else:
                r, g, b, a = BG_R, BG_G, BG_B, 255

            raw += struct.pack("BBBB", r, g, b, a)

    def chunk(chunk_type: bytes, data: bytes) -> bytes:
        c = chunk_type + data
        crc = struct.pack(">I", zlib.crc32(c) & 0xFFFFFFFF)
        return struct.pack(">I", len(data)) + c + crc

    # PNG signature
    sig = b"\x89PNG\r\n\x1a\n"

    # IHDR
    ihdr_data = struct.pack(">IIBBBBB", width, height, 8, 6, 0, 0, 0)
    ihdr = chunk(b"IHDR", ihdr_data)

    # IDAT
    compressed = zlib.compress(raw)
    idat = chunk(b"IDAT", compressed)

    # IEND
    iend = chunk(b"IEND", b"")

    return sig + ihdr + idat + iend


def main():
    os.makedirs(OUT_DIR, exist_ok=True)

    for size in [192, 512]:
        png = create_png(size, size)
        path = os.path.join(OUT_DIR, f"icon-{size}.png")
        with open(path, "wb") as f:
            f.write(png)
        print(f"  ✓ {path}  ({len(png):,} bytes)")

    print("\nDone. Icons generated.")


if __name__ == "__main__":
    main()

File: scripts/generate-icons.mjs
/**
 * Generate PWA icon PNGs (192x192 and 512x512) using Node.js stdlib only.
 * Usage: node scripts/generate-icons.mjs
 */
import { writeFileSync, mkdirSync } from "fs";
import { deflateSync, crc32 } from "zlib";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const OUT_DIR = resolve(__dirname, "..", "public");

mkdirSync(OUT_DIR, { recursive: true });

function createPng(width, height) {
  const BG_R = 0xf7, BG_G = 0xf3, BG_B = 0xea; // warm cream
  const ACC_R = 0x5d, ACC_G = 0x40, ACC_B = 0x37; // dark brown

  const cx = width / 2, cy = height / 2;
  const radius = width * 0.38;
  const radiusSq = radius * radius;

  // Build raw pixel rows (RGBA)
  const rawRows = [];
  for (let y = 0; y < height; y++) {
    const row = [0]; // filter byte: none
    for (let x = 0; x < width; x++) {
      const dx = x - cx, dy = y - cy;
      const inCircle = dx * dx + dy * dy <= radiusSq;
      if (inCircle) {
        row.push(ACC_R, ACC_G, ACC_B, 255);
      } else {
        row.push(BG_R, BG_G, BG_B, 255);
      }
    }
    rawRows.push(Buffer.from(row));
  }

  const raw = Buffer.concat(rawRows);

  function chunk(type, data) {
    const len = Buffer.alloc(4);
    len.writeUInt32BE(data.length, 0);
    const typeAndData = Buffer.concat([Buffer.from(type, "ascii"), data]);
    const crc = Buffer.alloc(4);
    crc.writeUInt32BE(crc32(typeAndData) >>> 0, 0);
    return Buffer.concat([len, typeAndData, crc]);
  }

  // IHDR
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8;  // bit depth
  ihdr[9] = 6;  // color type RGBA
  ihdr[10] = 0; // compression
  ihdr[11] = 0; // filter
  ihdr[12] = 0; // interlace

  const compressed = deflateSync(raw);

  return Buffer.concat([
    Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]), // PNG signature
    chunk("IHDR", ihdr),
    chunk("IDAT", compressed),
    chunk("IEND", Buffer.alloc(0)),
  ]);
}

for (const size of [192, 512]) {
  const png = createPng(size, size);
  const path = resolve(OUT_DIR, `icon-${size}.png`);
  writeFileSync(path, png);
  console.log(`  ✓ ${path}  (${png.length.toLocaleString()} bytes)`);
}

console.log("\nDone. Icons generated.");
